src/components/Identity/util.js

```diff
@@ -0,0 +1,61 @@
+import { AUTH_STATES } from "./constants";
+import isObject from "../../utils/isObject";
+
+const ERROR_MESSAGE = "Invalid customer Id format.";
+const NOT_AN_OBJECT_ERROR = "Each namespace should be an object.";
+const NO_ID_ERROR = "Each namespace object should have an id.";
+
+const validateCustomerIds = customerIds => {
+  if (!isObject(customerIds)) {
+    throw new Error(`${ERROR_MESSAGE} ${NOT_AN_OBJECT_ERROR}`);
+  }
+  Object.keys(customerIds).forEach(customerId => {
+    if (!isObject(customerIds[customerId])) {
+      throw new Error(`${ERROR_MESSAGE} ${NOT_AN_OBJECT_ERROR}`);
+    }
+    if (!customerIds[customerId].id) {
+      throw new Error(`${ERROR_MESSAGE} ${NO_ID_ERROR}`);
+    }
+  });
+};
+
+const normalizeCustomerIds = customerIds => {
+  const normalized = {};
+  Object.keys(customerIds).forEach(customerId => {
+    const currentId = customerIds[customerId];
+    normalized[customerId] = { ...currentId };
+    if (
+      currentId.authState !== AUTH_STATES.AUTHENTICATED ||
+      currentId.authState !== AUTH_STATES.LOGGED_OUT
+    ) {
+      normalized[customerId].authState = AUTH_STATES.UNKNOWN;
+    }
+  });
+  return normalized;
+};
+
+const serializeCustomerIds = customerIds => {
+  return Object.keys(customerIds)
+    .reduce((serialized, customerId) => {
+      return `${serialized}${customerId}|${customerIds[customerId].id}${
+        customerIds[customerId].authState
+      }|`;
+    }, "")
+    .slice(0, -1);
+};
+
+const createHashFromString = string => {
+  return string.split('')
+    .map(char => char.charCodeAt(0))
+    .reduce((hash, charCode) => {
+      const shift = (hash << 5) - hash + charCode; // eslint-disable-line no-bitwise
+      return shift & shift; // eslint-disable-line no-bitwise
+    }, 0);
```

Comment:
This hashing algorithm is based on Java's hashCode algorithm, which isn't really a standard. It's probably worth using a better, more standard algorithm. CRC32, MD5, and SHA-1, while not cryptographically secure, are fine for comparing two inputs and would all result in a much smaller hash than SHA-256. While there's no native API for generating a CRC32, it would likely be the fastest, produce the smallest hash, and be the most appropriate for what we're trying to do here. It's probably the smallest amount of code too, even when taking into consideration that SHA-1 has a native API. There's a 1 in 4,294,967,296 chance of collision between two inputs with CRC32, which should be sufficient. I believe this is a proper implementation of CRC32: http://crc32.nichabi.com/javascript-function.php

Reply:
You probably want to run some UTF-8 strings through it and compare the output to some online CRC32 hashing tools.

Reply:
CRC32 returns a number. Since cookies are stored as strings, each digit in the CRC23 number is represented as one character. You can shorten the number of character needed by converting the number to a string that takes advantage of characters other than 0-9. So if you take the output of CRC32 and call `.toString(36)` you'll get a string using characters 0-9 and A-Z, resulting in a string that's about 6 or 7 characters long instead of a 9 or 10 characters. Technically, we could use an even higher radix by leveraging lowercase characters, etc., but there's no native API for that so it's probably not worth the code it would take to implement it.